File: Cpp/472.cpp
// Given a list of words (without duplicates), please write a program that
// returns all concatenated words in the given list of words. A concatenated
// word is defined as a string that is comprised entirely of at least two
// shorter words in the given array.
//
// Example:
// Input:
// ["cat","cats","catsdogcats","dog","dogcatsdog","hippopotamuses","rat","ratcatdogcat"]
//
// Output: ["catsdogcats","dogcatsdog","ratcatdogcat"]
//
// Explanation: "catsdogcats" can be concatenated by "cats", "dog" and "cats";
//  "dogcatsdog" can be concatenated by "dog", "cats" and "dog";
// "ratcatdogcat" can be concatenated by "rat", "cat", "dog" and "cat".
// Note:
// The number of elements of the given array will not exceed 10,000
// The length sum of elements in the given array will not exceed 600,000.
// All the input string will only include lower case letters.
// The returned elements order does not matter.

class Solution {
public:
    vector<string> findAllConcatenatedWordsInADict(vector<string>& words) {
        unordered_set<string> words_set(words.begin(), words.end());
        vector<string> res;

        for (int k = 0; k < words.size(); ++k) {
            int n = words[k].size();
            vector<int> dp(n + 1, 0);
            dp[0] = 1;

            for (int i = 1; i <= n; ++i) {
                for (int j = i - 1; j >= 0; --j) {
                    if (dp[j]) {
                        if (i - j < n
                            && words_set.count(words[k].substr(j, i - j))) {
                            dp[i] = 1;
                            break;
                        }
                    }
                }
            }

            if (n != 0 && dp[n]) {
                res.push_back(words[k]);
            }
        }

        return res;
    }
};

File: Python/472.py
DFS
it usually takes 300 ms

class Solution(object):
    def findAllConcatenatedWordsInADict(self, words):
        """
        :type words: List[str]
        :rtype: List[str]
        """
        ws = set()
        words = sorted(words, key= lambda w:len(w))
        res = []
        def dfs(w):
            if w in ws:
                return True
            for i in range(1, len(w)):
                if w[:i] in ws and dfs(w[i:]):
                    return True
            return False
        
        for k in range(len(words)):
            if dfs(words[k]):
                res.append(words[k])
            ws.add(words[k])
        return res
DP
it usually takes 2000ms

class Solution(object):
    def findAllConcatenatedWordsInADict(self, words):
        """
        :type words: List[str]
        :rtype: List[str]
        """
        ws = set()
        words = sorted(words, key= lambda w:len(w))
        res = []
        for k in range(len(words)):
            n = len(words[k])
            dp = [False]*(n+1)
            dp[0] = True
            
            for i in range(1, n+1):
                for j in range(i-1, -1, -1):
                    if dp[j] and (i-j < n) and words[k][j:i] in ws:
                        dp[i] = True
                        break
            if n!=0 and dp[n]:
                res.append(words[k])
            ws.add(words[k])
        return res
